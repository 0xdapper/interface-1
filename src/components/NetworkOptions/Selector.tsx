import ellipsisLogo from 'assets/svg/ellipsis.svg'
import Row from 'components/Row'
import { getChainInfo } from 'constants/chainInfo'
import { ChevronDown, ChevronUp } from 'react-feather'
import styled from 'styled-components'

const StyledButton = styled.button<{ isActive: boolean; showChevron?: boolean }>`
  height: 100%;
  margin: 0;
  padding: 6px ${({ showChevron }) => (showChevron ? '4px' : '10px')} 6px 12px;
  border-radius: 12px;
  border: none;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
  background-color: ${({ theme }) => theme.backgroundInteractive};
  overflow: hidden;

  &::before {
    content: '';
    position: absolute;
    border-radius: inherit;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: ${({ isActive, theme }) => (isActive ? theme.stateOverlayPressed : theme.stateOverlayHover)};
    pointer-events: none;
    opacity: ${({ isActive }) => (isActive ? 1 : 0)};
  }

  &:hover::before {
    opacity: ${({ isActive }) => (isActive ? 0 : 1)};
  }
`
const SquareChainLogoContainer = styled.div<{ isActive?: boolean }>`
  border-radius: 8px;
  width: 24px;
  height: 24px;
  outline: 2px solid ${({ theme }) => theme.backgroundInteractive};
  margin-left: -2px;
`
export const SquareChainLogo = styled.img`
  height: 100%;
  width: 100%;
  background: #505a78;
  border-radius: 6px;
  width: 24px;
  height: 24px;
`
const Chevron = styled.span<{ isActive: boolean }>`
  padding-top: 1px;
  color: ${({ theme }) => theme.textSecondary};
`
function AllNetworks({ chainIds, isActive }: { chainIds: number[]; isActive: boolean }) {
  return (
    <Row data-testid="all-networks-selected">
      {chainIds.slice(0, 3).map((chainId) => {
        const chainInfo = getChainInfo(chainId)
        return (
          <SquareChainLogoContainer key={chainId} isActive={isActive}>
            <SquareChainLogo src={chainInfo?.squareLogoUrl} alt={chainInfo?.label} />
          </SquareChainLogoContainer>
        )
      })}
      <SquareChainLogoContainer isActive={isActive}>
        <SquareChainLogo src={ellipsisLogo} alt="More network options" />
      </SquareChainLogoContainer>
    </Row>
  )
}
export function Selector({
  chainIds,
  selectedChainId = 0,
  isActive,
  onClick,
}: {
  chainIds: number[]
  selectedChainId?: number
  isActive: boolean
  onClick: () => void
}) {
  const chainInfo = getChainInfo(selectedChainId)
  return (
    <StyledButton
      isActive={isActive}
      showChevron={!!selectedChainId}
      onClick={onClick}
      aria-label="Toggle Network Options"
      data-testid="network-selector"
    >
      {selectedChainId === 0 ? (
        <AllNetworks chainIds={chainIds} isActive={isActive} />
      ) : (
        <>
          <SquareChainLogoContainer key={selectedChainId} isActive={isActive}>
            <SquareChainLogo
              src={chainInfo?.squareLogoUrl}
              alt={chainInfo?.label}
              data-testid={`network-${selectedChainId}-selected`}
            />
          </SquareChainLogoContainer>
          <Chevron isActive={isActive}>
            {isActive ? (
              <ChevronUp width={20} height={15} viewBox="0 0 24 20" data-testid="network-options-chevron-up" />
            ) : (
              <ChevronDown width={20} height={15} viewBox="0 0 24 20" data-testid="network-options-chevron-down" />
            )}
          </Chevron>
        </>
      )}
    </StyledButton>
  )
}
