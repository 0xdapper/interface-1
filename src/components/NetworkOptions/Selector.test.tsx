import userEvent from '@testing-library/user-event'
import { act, render, screen } from 'test-utils/render'
import noop from 'utils/noop'

import { Selector } from './Selector'

describe('button for selecting a network filter', () => {
  const chainIds = [1, 10, 137] // ETH, OP, POLYGON

  it('defaults to showing all networks with logo pile', () => {
    const { container } = render(<Selector chainIds={chainIds} isActive={false} onClick={noop} />)
    expect(container).toMatchSnapshot()
  })

  it('shows single network logo and chevron when selected', () => {
    const { container } = render(<Selector chainIds={chainIds} selectedChainId={1} isActive={false} onClick={noop} />)
    expect(container).toMatchSnapshot()
  })

  describe('non-idle styling', () => {
    it('changes when hovered', async () => {
      const { container } = render(<Selector chainIds={chainIds} isActive={false} onClick={noop} />)
      await act(() => userEvent.hover(screen.getByTestId('network-selector')))
      expect(container).toMatchSnapshot()
    })

    it('changes when active and hovered', async () => {
      const { container } = render(<Selector chainIds={chainIds} isActive={true} onClick={noop} />)
      await act(() => userEvent.hover(screen.getByTestId('network-selector')))
      expect(container).toMatchSnapshot()
    })
  })
})
