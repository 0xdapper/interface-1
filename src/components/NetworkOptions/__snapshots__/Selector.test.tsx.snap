// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button for selecting a network filter defaults to showing all networks with logo pile 1`] = `
.c1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c0 {
  height: 100%;
  margin: 0;
  padding: 6px 10px 6px 12px;
  border-radius: 12px;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  background-color: #E8ECFB;
  overflow: hidden;
}

.c0::before {
  content: '';
  position: absolute;
  border-radius: inherit;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: #98A1C014;
  pointer-events: none;
  opacity: 0;
}

.c0:hover::before {
  opacity: 1;
}

.c3 {
  border-radius: 8px;
  width: 24px;
  height: 24px;
  outline: 2px solid #E8ECFB;
  margin-left: -2px;
}

.c4 {
  height: 100%;
  width: 100%;
  background: #505a78;
  border-radius: 6px;
  width: 24px;
  height: 24px;
}

<div>
  <button
    aria-label="Toggle Network Options"
    class="c0"
    data-testid="network-selector"
  >
    <div
      class="c1 c2"
      data-testid="all-networks-selected"
    >
      <div
        class="c3"
      >
        <img
          alt="Ethereum"
          class="c4"
          src="ethereum_square_logo.svg"
        />
      </div>
      <div
        class="c3"
      >
        <img
          alt="Optimism"
          class="c4"
          src="optimism_square_logo.svg"
        />
      </div>
      <div
        class="c3"
      >
        <img
          alt="Polygon"
          class="c4"
          src="polygon_square_logo.svg"
        />
      </div>
      <div
        class="c3"
      >
        <img
          alt="More network options"
          class="c4"
          src="ellipsis.svg"
        />
      </div>
    </div>
  </button>
</div>
`;

exports[`button for selecting a network filter non-idle styling changes when active and hovered 1`] = `
.c1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c0 {
  height: 100%;
  margin: 0;
  padding: 6px 10px 6px 12px;
  border-radius: 12px;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  background-color: #E8ECFB;
  overflow: hidden;
}

.c0::before {
  content: '';
  position: absolute;
  border-radius: inherit;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: #B8C0DC3d;
  pointer-events: none;
  opacity: 1;
}

.c0:hover::before {
  opacity: 0;
}

.c3 {
  border-radius: 8px;
  width: 24px;
  height: 24px;
  outline: 2px solid #E8ECFB;
  margin-left: -2px;
}

.c4 {
  height: 100%;
  width: 100%;
  background: #505a78;
  border-radius: 6px;
  width: 24px;
  height: 24px;
}

<div>
  <button
    aria-label="Toggle Network Options"
    class="c0"
    data-testid="network-selector"
  >
    <div
      class="c1 c2"
      data-testid="all-networks-selected"
    >
      <div
        class="c3"
      >
        <img
          alt="Ethereum"
          class="c4"
          src="ethereum_square_logo.svg"
        />
      </div>
      <div
        class="c3"
      >
        <img
          alt="Optimism"
          class="c4"
          src="optimism_square_logo.svg"
        />
      </div>
      <div
        class="c3"
      >
        <img
          alt="Polygon"
          class="c4"
          src="polygon_square_logo.svg"
        />
      </div>
      <div
        class="c3"
      >
        <img
          alt="More network options"
          class="c4"
          src="ellipsis.svg"
        />
      </div>
    </div>
  </button>
</div>
`;

exports[`button for selecting a network filter non-idle styling changes when hovered 1`] = `
.c1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c0 {
  height: 100%;
  margin: 0;
  padding: 6px 10px 6px 12px;
  border-radius: 12px;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  background-color: #E8ECFB;
  overflow: hidden;
}

.c0::before {
  content: '';
  position: absolute;
  border-radius: inherit;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: #98A1C014;
  pointer-events: none;
  opacity: 0;
}

.c0:hover::before {
  opacity: 1;
}

.c3 {
  border-radius: 8px;
  width: 24px;
  height: 24px;
  outline: 2px solid #E8ECFB;
  margin-left: -2px;
}

.c4 {
  height: 100%;
  width: 100%;
  background: #505a78;
  border-radius: 6px;
  width: 24px;
  height: 24px;
}

<div>
  <button
    aria-label="Toggle Network Options"
    class="c0"
    data-testid="network-selector"
  >
    <div
      class="c1 c2"
      data-testid="all-networks-selected"
    >
      <div
        class="c3"
      >
        <img
          alt="Ethereum"
          class="c4"
          src="ethereum_square_logo.svg"
        />
      </div>
      <div
        class="c3"
      >
        <img
          alt="Optimism"
          class="c4"
          src="optimism_square_logo.svg"
        />
      </div>
      <div
        class="c3"
      >
        <img
          alt="Polygon"
          class="c4"
          src="polygon_square_logo.svg"
        />
      </div>
      <div
        class="c3"
      >
        <img
          alt="More network options"
          class="c4"
          src="ellipsis.svg"
        />
      </div>
    </div>
  </button>
</div>
`;

exports[`button for selecting a network filter shows single network logo and chevron when selected 1`] = `
.c0 {
  height: 100%;
  margin: 0;
  padding: 6px 4px 6px 12px;
  border-radius: 12px;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  background-color: #E8ECFB;
  overflow: hidden;
}

.c0::before {
  content: '';
  position: absolute;
  border-radius: inherit;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: #98A1C014;
  pointer-events: none;
  opacity: 0;
}

.c0:hover::before {
  opacity: 1;
}

.c1 {
  border-radius: 8px;
  width: 24px;
  height: 24px;
  outline: 2px solid #E8ECFB;
  margin-left: -2px;
}

.c2 {
  height: 100%;
  width: 100%;
  background: #505a78;
  border-radius: 6px;
  width: 24px;
  height: 24px;
}

.c3 {
  padding-top: 1px;
  color: #7780A0;
}

<div>
  <button
    aria-label="Toggle Network Options"
    class="c0"
    data-testid="network-selector"
  >
    <div
      class="c1"
    >
      <img
        alt="Ethereum"
        class="c2"
        data-testid="network-1-selected"
        src="ethereum_square_logo.svg"
      />
    </div>
    <span
      class="c3"
    >
      <svg
        data-testid="network-options-chevron-down"
        fill="none"
        height="15"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="6 9 12 15 18 9"
        />
      </svg>
    </span>
  </button>
</div>
`;
